File: telegram-bot/src/handlers/rate.js
const db = require('../db');
const pendingRate = require('../pendingRate');
const { code } = require('../format');

async function rateHandler(ctx) {
  db.ensureUser(ctx.from.id);
  const preferred = db.getPreferences(ctx.from.id);

  if (preferred.length === 0) {
    return ctx.replyWithMarkdownV2('You have not picked any preferred currencies yet\\. Use /setpreferred first\\.');
  }

  pendingRate.mark(ctx.from.id);
  const list = preferred.map((c) => code(c.toUpperCase())).join(', ');
  await ctx.replyWithMarkdownV2(
    `*Your preferred currencies:*\n${list}\n\nSend me the base currency you would like to compare against \\(e\\.g\\. ${code(
      'USD'
    )}\\)\\.`
  );
}

module.exports = rateHandler;

File: telegram-bot/src/handlers/rate.py
from .. import db, pending_rate
from ..format import escape_md
from ..reply import send_rich_message

PROMPT = 'Send me the base currency you would like to compare against (e.g. USD).'


def build_rate_view(preferred):
    codes = [c.upper() for c in preferred]
    rich = {
        'markdown': (
            '# Your preferred currencies\n\n'
            + '\n'.join(f'- {escape_md(c)}' for c in codes)
            + f'\n\n{PROMPT}'
        ),
        'fallback': f"Your preferred currencies:\n{', '.join(codes)}\n\n{PROMPT}",
    }
    return rich, None


async def rate_handler(event):
    db.ensure_user(event.sender_id)
    preferred = db.get_preferences(event.sender_id)

    if not preferred:
        return await event.respond(
            'You have not picked any preferred currencies yet. Use /setpreferred first.',
            parse_mode='md',
        )

    pending_rate.mark(event.sender_id)
    rich, buttons = build_rate_view(preferred)
    await send_rich_message(event.client, event.chat_id, rich, buttons)
